numpy/random/_generator.pyx

```diff
@@ -4162,7 +4216,127 @@ cdef class Generator:
+        --------
+        numpy.random.Generator.randomly_permuted
+        numpy.random.Generator.shuffle
+        numpy.random.Generator.permutation
+
+        Examples
+        --------
+        Create a `numpy.random.Generator` instance:
+
+        >>> rng = np.random.Generator(np.random.PCG64(12345))
+
+        Create a test array:
+
+        >>> x = np.arange(24).reshape(3, 8)
+        >>> x
+        array([[ 0,  1,  2,  3,  4,  5,  6,  7],
+               [ 8,  9, 10, 11, 12, 13, 14, 15],
+               [16, 17, 18, 19, 20, 21, 22, 23]])
+
+        Shuffle the rows of `x` in-place:
+
+        >>> rng.randomly_permute(x, axis=1)
+        >>> x
+        array([[ 4,  3,  0,  2,  1,  6,  7,  5],
+               [10, 12, 14, 15,  9, 11,  8, 13],
+               [16, 20, 19, 21, 18, 22, 17, 23]])
+
+        """
+        cdef int ax
+        cdef np.npy_intp axlen, axstride, itemsize
+        cdef void *buf
+        cdef np.flatiter it
+
+        if type(x) is not np.ndarray:
+            raise ValueError('x must be a numpy array')
+        if axis is None:
+            self.shuffle(x.flat)
+            return
+        ax = normalize_axis_index(axis, np.ndim(x))
+        itemsize = x.itemsize
+        axlen = x.shape[ax]
+        axstride = x.strides[ax]
+
+        it = np.PyArray_IterAllButAxis(x, &ax)
+
+        buf = malloc(itemsize)
+        if buf == NULL:
+            raise MemoryError('memory allocation failed in randomly_permute')
+
+        with self.lock, nogil:
+            while np.PyArray_ITER_NOTDONE(it):
+                _shuffle_raw(&self._bitgen, axlen, 0, itemsize, axstride,
+                             <char *>np.PyArray_ITER_DATA(it), <char *>buf)
+                np.PyArray_ITER_NEXT(it)
+
+        free(buf)
+
+    def randomly_permuted(self, object x, axis=None):
+        """
+        Randomly shuffle slices along an axis.
+
+        If `axis` None, the flattened array is shuffled.
+
+        Parameters
+        ----------
+        x : ndarray
+            Array to be randomly permuted.
+        axis : int, optional
+            Slices of `x` in this axis are permuted. Each slice
+            is permuted independently of the others.
+
+        Returns
+        -------
+        A randomly permuted copy of `x`.
+
+        See Also
+        --------
+        numpy.random.Generator.randomly_permute
+        numpy.random.Generator.shuffle
+        numpy.random.Generator.permutation
+
+        Examples
+        --------
+        >>> rng = np.random.Generator(np.random.PCG64(12345))
+        >>> x = np.arange(24).reshape(3, 8)
+        >>> x
+        array([[ 0,  1,  2,  3,  4,  5,  6,  7],
+               [ 8,  9, 10, 11, 12, 13, 14, 15],
+               [16, 17, 18, 19, 20, 21, 22, 23]])
+        >>> rng.randomly_permuted(x, axis=1)
+        array([[ 4,  3,  0,  2,  1,  6,  7,  5],
+               [10, 12, 14, 15,  9, 11,  8, 13],
+               [16, 20, 19, 21, 18, 22, 17, 23]])
+
+        """
+        y = np.copy(x)
```

Comment:
Having an entire second function to save 10 characters seems excessive. The inline version seems to be more useful since endusers can always create the version by taking a copy themselves. 

These also have difference types accepted.  permuted takes array_like while permute takes only ndarray.

Reply:
The in-place version requires an `ndarray` instance, so it doesn't have to reproduce the shape discovery functionality of `numpy.array`, and it doesn't have to deal with encountering immutable elements somewhere in the depths of the object (e.g. `[[(1, 2, 3), (4, 5, 6)], [(2, 4, 6), (1, 2, 3)]]`), nor ragged arrays (e.g. `[[1, 2, 3], [4, 5], [6, 7, 8]]`), etc.  Lately I've found it useful to think of the in-place version as an `ndarray` method rather than a function that accepts an array-like, analogous to the `sort` method.